upload-api/tables/revocations.js

```diff
@@ -0,0 +1,87 @@
+import {
+  DynamoDBClient,
+  UpdateItemCommand,
+  BatchGetItemCommand
+} from '@aws-sdk/client-dynamodb'
+import { marshall, unmarshall } from '@aws-sdk/util-dynamodb'
+import { parseLink } from '@ucanto/core'
+
+/**
+ * Abstraction layer to handle operations on revocations table.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {object} [options]
+ * @param {string} [options.endpoint]
+ */
+export function createRevocationsTable(region, tableName, options = {}) {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint,
+  })
+
+  return useRevocationsTable(dynamoDb, tableName)
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} tableName
+ * @returns {import('../types').RevocationsTable}
+ */
+export function useRevocationsTable(dynamoDb, tableName) {
+  return {
+    async put (delegationCID, contextCID, causeCID) {
+      await dynamoDb.send(new UpdateItemCommand({
+        TableName: tableName,
+        Key: marshall({
+          delegation: delegationCID.toString(),
+        }),
+        UpdateExpression: 'ADD contextsAndCauses :candc',
+        ExpressionAttributeValues: marshall({
+          ':candc': new Set([`${contextCID.toString()}:${causeCID.toString()}`])
+        })
+      }))
+    },
+
+    async getRevocations(delegationCIDs) {
+      // BatchGetItem only supports batches of 100 and return values under 16MB
+      // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchGetItem.html
+      // limiting to 100 should be fine for now, and since we don't return much data we should always
+      // stay under the 16 MB limit
+      if (delegationCIDs.length > 100) {
+        throw new Error('checking for more than 100 revocations in a single call is currently not supported')
+      }
```

Comment:
I think we should do multiple queries here as opposed to failing, caller will not have a way of handle this other than passing less CIDs and then doing another query.

I can't believe I'm saying this but perhaps return value of this function should be `AsyncIterable<{revoke: UCANLink, scope: UCANLink, cause:UCANLink}[]>` that way caller could check first page of scopes and if invocation still valid, get another page and etc... Which I think it would be better API as it does not require you to handle errors and instead just does query at time.

Reply:
Also please do not throw, just return Result type so errors could be typed.

Reply:
yeaaaa I just figured we won't see this case for a while - if someone is sending us a 100+ item delegation chain I'm pretty comfortable just throwing an error for now....

re: `Result` this is the pattern we tend to use for `w3infra`-internal tables (ie, tables that are not part of the interface between `w3up` and `w3infra`):

https://github.com/web3-storage/w3infra/blob/main/upload-api/tables/consumer.js
https://github.com/web3-storage/w3infra/blob/main/upload-api/tables/store.js
https://github.com/web3-storage/w3infra/blob/main/upload-api/tables/subscription.js
https://github.com/web3-storage/w3infra/blob/main/upload-api/tables/upload.js

so I'd prefer to stick with this style for consistency. I'm not opposed to switching to using `Result` for these tables, but it feels like it's worth a broader conversation and a concerted effort to convert them all if we do go that way... what do you think?

Reply:
adding - with the new interface I think this will actually be a standalone store instead of an internal table, so I'll use `Result` after all!

Reply:
done